internal/controller/multiclusterservice_controller.go

```diff
@@ -192,40 +198,133 @@ func (r *MultiClusterServiceReconciler) updateStatus(ctx context.Context, mcs *k
 	return nil
 }
 
+// setClustersServicesReadinessConditions calculates and sets
+// [github.com/K0rdent/kcm/api/v1alpha1.ServicesInReadyStateCondition] and
+// [github.com/K0rdent/kcm/api/v1alpha1.ClusterInReadyStateCondition]
+// informational conditions with the number of ready services and clusters.
+func (r *MultiClusterServiceReconciler) setClustersServicesReadinessConditions(ctx context.Context, mcs *kcm.MultiClusterService) error {
+	sel, err := metav1.LabelSelectorAsSelector(&mcs.Spec.ClusterSelector)
+	if err != nil {
+		return fmt.Errorf("failed to construct selector from MultiClusterService %s selector: %w", client.ObjectKeyFromObject(mcs), err)
+	}
+
+	clusters := &metav1.PartialObjectMetadataList{}
+	clusters.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "cluster.x-k8s.io",
+		Version: "v1beta1",
+		Kind:    "Cluster",
+	})
+	if err := r.Client.List(ctx, clusters, client.MatchingLabelsSelector{Selector: sel}); err != nil {
+		return fmt.Errorf("failed to list partial Clusters: %w", err)
+	}
+
+	ready := 0
+	for _, cluster := range clusters.Items {
+		key := client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Name}
+		cld := new(kcm.ClusterDeployment)
+		if err := r.Client.Get(ctx, key, cld); err != nil {
+			return fmt.Errorf("failed to get ClusterDeployment %s: %w", key.String(), err)
+		}
+
+		rc := apimeta.FindStatusCondition(cld.Status.Conditions, kcm.ReadyCondition)
+		if rc != nil && rc.Status == metav1.ConditionTrue {
+			ready++
+		}
+	}
+
+	desiredClusters, desiredServices := len(clusters.Items), len(clusters.Items)*len(mcs.Spec.ServiceSpec.Services)
+	c := metav1.Condition{
+		Type:    kcm.ClusterInReadyStateCondition,
+		Status:  metav1.ConditionTrue,
+		Reason:  kcm.SucceededReason,
+		Message: fmt.Sprintf("%d/%d", ready, desiredClusters),
+	}
+	if ready != desiredClusters {
+		c.Reason = kcm.ProgressingReason
+		c.Status = metav1.ConditionFalse
+	}
+
+	apimeta.SetStatusCondition(&mcs.Status.Conditions, c)
+	apimeta.SetStatusCondition(&mcs.Status.Conditions, getServicesReadinessCondition(mcs.Status.Services, desiredServices))
+
+	return nil
+}
+
+func getServicesReadinessCondition(serviceStatuses []kcm.ServiceStatus, desiredServices int) metav1.Condition {
+	ready := 0
+	for _, svcstatus := range serviceStatuses {
+		svcready := true
+		if slices.ContainsFunc(svcstatus.Conditions, func(e metav1.Condition) bool { return e.Status != metav1.ConditionTrue }) {
+			svcready = false
+		}
+
+		if svcready {
+			ready++
+		}
```

Comment:
Can it be simplified a bit?
```suggestion
		if !slices.ContainsFunc(svcstatus.Conditions, func(e metav1.Condition) bool { return e.Status != metav1.ConditionTrue }) {
			ready++
		}
```

Reply:
it can. thanks!